core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java

```diff
@@ -326,4 +327,14 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeBoolean(updateAllTypes);
         out.writeOptionalWriteable(concreteIndex);
     }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        if (source != null) {
+            builder.rawValue(new BytesArray(source), XContentType.JSON);
+        } else {
+            builder.startObject().endObject();
```

Comment:
is it ok here to only end the object? should we not start it too? and above, does rawValue print out a valid complete object?

Reply:
I don't understand the first question; we are both starting and ending it, aren't we?

And yes, as `source` has to be a complete object, `rawValue` applied by itself to the builder also produces one.